contracts/test/LONStaking.t.sol

```diff
@@ -433,80 +471,302 @@ contract LONStakingTest is Test {
 
-        uint256 redeemAmount = lonStaking.balanceOf(user).div(2);
         uint256 expectedLONAmount = _getExpectedLONWithoutPenalty(redeemAmount);
+        vm.prank(redeemer);
         lonStaking.redeem(redeemAmount);
-        vm.stopPrank();
 
         uint256 lonStakingXLONAfter = lonStaking.totalSupply();
         assertEq(lonStakingXLONAfter, lonStakingXLONBefore - redeemAmount);
         lonStakingLON.assertChange(-int256(expectedLONAmount));
-        userXLON.assertChange(-int256(redeemAmount));
-        userLON.assertChange(int256(expectedLONAmount));
-        assertGt(lonStaking.stakersCooldowns(user), 0, "Cooldown record remains until user redeem all shares");
+        redeemerXLON.assertChange(-int256(redeemAmount));
+        redeemerLON.assertChange(int256(expectedLONAmount));
+        if (redeemPartial) {
+            assertGt(lonStaking.stakersCooldowns(redeemer), 0, "Cooldown record remains until user redeem all shares");
+        }
+    }
+
+    function testRedeemPartial() public {
+        _stake(user, DEFAULT_STAKE_AMOUNT);
+        vm.prank(user);
+        lonStaking.unstake();
+
+        vm.warp(block.timestamp + COOLDOWN_SECONDS + 1);
+
+        uint256 redeemAmount = lonStaking.balanceOf(user).div(2);
+        _redeemAndValidate(user, redeemAmount);
+    }
+
+    function testFuzz_RedeemPartial(uint256 stakeAmount, uint256 redeemAmount) public {
+        vm.assume(stakeAmount > 0);
+        vm.assume(redeemAmount > 0);
+        vm.assume(redeemAmount < stakeAmount);
+        vm.assume(stakeAmount <= lon.cap());
+        vm.assume(stakeAmount.add(lon.totalSupply()) <= lon.cap());
+
+        _stake(user, stakeAmount);
+        vm.prank(user);
+        lonStaking.unstake();
+
+        vm.warp(block.timestamp + COOLDOWN_SECONDS + 1);
+
+        _redeemAndValidate(user, redeemAmount);
+    }
+
+    function _determineStakeRedeemAmount(uint80[2][16] memory stakeAndRedeemAmounts)
+        internal
+        pure
+        returns (uint256[16] memory stakeAmounts, uint256[16] memory redeemAmounts)
+    {
+        for (uint256 i = 0; i < stakeAndRedeemAmounts.length; i++) {
+            if (stakeAndRedeemAmounts[i][0] >= stakeAndRedeemAmounts[i][1]) {
+                stakeAmounts[i] = stakeAndRedeemAmounts[i][0];
+                redeemAmounts[i] = stakeAndRedeemAmounts[i][1];
+            } else {
+                stakeAmounts[i] = stakeAndRedeemAmounts[i][1];
+                redeemAmounts[i] = stakeAndRedeemAmounts[i][0];
+            }
+        }
+    }
+
+    function testFuzz_RedeemPartialMultiple_OneByOne(uint80[2][16] memory stakeAndRedeemAmounts) public {
+        // LON cap lies between 2**87 and 2**88, setting upper bound of stake amount to 2**80 - 1 so stake amount will not exceed cap
+        uint256 totalLONAmount = lon.totalSupply();
+        (uint256[16] memory stakeAmounts, uint256[16] memory redeemAmounts) = _determineStakeRedeemAmount(stakeAndRedeemAmounts);
```

Comment:
Technically there are changes where `stakeAmount == redeemAmount` so it's not exactly partial redeem in every cases.

Reply:
We can check `stakeAmount > redeemAmount` but again it will take many time regenerating the whole 16 numbers.